### ISSUE TITLE
dials.refine: filter out poor centroids at scan edges

### ISSUE THREAD

**Issue by dagewa** [dials/dials#457]
Add something like a "trim" parameter to `dials.refine` to filter out centroids within, say, 0.5-1.0 images from the edges of scans as these have the worst phi centroids. Does this improve integration results for narrow wedge datasets? Takanori observes that DIALS struggles with small wedges (lower index rate, higher RMSD, poorer CC1/2). Here are links to data sets to check:

> https://zenodo.org/record/887440
> This is the raw diffraction images for the structure of a GPCR, LPA6.
> It consists of hundreds of small wedge (- 10 deg) sweeps.
> 
> https://zenodo.org/record/897676
> This is ETBR, another GPCR. It consists of fewer sweeps.
> 
> Raw data for https://www.nature.com/articles/s41477-017-0022-8 will also
> be uploaded
> soon.

**Comment by dagewa** [comment]
The raw data of TPT (small wedge, fine sliced) are also available. They are very large though.
https://data.sbgrid.org/dataset/511/ (303G)
https://data.sbgrid.org/dataset/512/ (107G)

**Comment by dagewa** [comment]
Hi Takanori,

I've just started looking at one 6 image dataset from the first tarball. Already things seem rather wrong:
```
dials.import ../5xsz_lpa6_unipuck1/16/multi001_master.h5
dials.find_spots datablock.json nproc=8
dials.index datablock.json strong.pickle unit_cell=55.9,65.0,160.7,90,90,90
dials.refine experiments.json indexed.pickle
dials.report refined_experiments.json refined.pickle
```
This gives a sawtooth plot in the difference between observed and calculated centroids. I don't yet underrstand why. I'll continue looking at this next week.
![sawtooth](https://user-images.githubusercontent.com/11502083/42102156-0591ef4c-7bbd-11e8-9654-e7cfbe9c3f8e.png)

**Comment by biochem-fan** [comment]
Thank you very much for investigation. Since some crystals are worse than others, could you please look at a few more crystals if this is general?

**Comment by dagewa** [comment]
No problem, I'll look at various datasets next week, until we work out what is going on.

**Comment by keitaroyam** [comment]
Dear David,

The dataset you are looking at (5xsz_lpa6_unipuck1/16/multi001_master.h5) was indexed by XDS with the correct unit cell parameters (without giving it) but there were some unindexed reflections. The following are probably some "good" datasets included in final merged data:
- 5xsz_lpa6_unipuck2/07/multi002_master.h5
- 5xsz_lpa6_unipuck2/09/multi004_master.h5
- 5xsz_lpa6_unipuck3/02data/multi001_master.h5

**Comment by dagewa** [comment]
Thanks for pointing out better quality datasets to me. I haven't downloaded unipuck3, but the strange sawtooth pattern is present for the other two. For example, for 5xsz_lpa6_unipuck2/07/multi002_master.h5:
![sawtooth2](https://user-images.githubusercontent.com/11502083/42156473-110ff3dc-7de3-11e8-8cde-f43725f5aefe.png)
I suspect this is exactly due to the effect I am hoping to remove by trimming centroids with Z values at 0.5 images from the scan edges. I'll test that and report back here.

**Comment by dagewa** [comment]
Ok, so interestingly keeping only reflections with observed 0.51 < Z centroids < 3.49 does not solve the problem. Doing that removed 37% of the total number of reflections, those with the worst centroids. However, the resulting plot after refinement still shows a sawtooth.
![sawtooth3](https://user-images.githubusercontent.com/11502083/42157279-3217f0f0-7de5-11e8-8178-277d7027d2fe.png)

**Comment by dagewa** [comment]
A simple scatterplot of Zdiff versus Zobs shows some weird features too
```
from dials.array_family import flex
from matplotlib import pyplot
rt=flex.reflection_table.from_pickle('indexed.pickle')
xo,yo,zo=rt['xyzobs.px.value'].parts()
xc,yc,zc=rt['xyzcal.px'].parts()
zd = zo-zc
pyplot.scatter(zo, zd)
pyplot.xlabel('Zobs')
pyplot.ylabel('Zobs - Zcalc')
pyplot.savefig('scatter.png')
```
![scatter](https://user-images.githubusercontent.com/11502083/42158333-060ab152-7de8-11e8-84c8-2ba3894c67ac.png)

**Comment by dagewa** [comment]
Helps to remove the reflections indexed (0,0,0)
![scatter](https://user-images.githubusercontent.com/11502083/42158480-73c3ff64-7de8-11e8-91d2-b1b0523e4ac8.png)
Anyway, I see there are always sharp features at every multiple of 0.5 images, which is where the Z centroids of spots that only appear on a single image are placed.

**Comment by dagewa** [comment]
Now removing all reflections at image centres:
```
from dials.array_family import flex
rt=flex.reflection_table.from_pickle('indexed.pickle')

# remove unindexed reflections
rt=rt.select(rt['miller_index'] != (0,0,0))

# remove reflections centred at each image
xo,yo,zo=rt['xyzobs.px.value'].parts()
to_keep=flex.bool(len(rt), True)
for i in range(int(max(zo)) + 1):
  im_centre = i + 0.5
  sel = (zo < im_centre - 0.001) | (zo > im_centre + 0.001)
  to_keep = to_keep & sel
print "removing {0}/{1} reflections".format(to_keep.count(False), len(rt))
rt = rt.select(to_keep)

# save filtered reflections
rt.as_pickle('indexed_filtered.pickle')
```
This removes 707 of 828 indexed reflections, so there's not much left to work with! Do refinement anyway:
```
dials.refine experiments.json indexed_filtered.pickle
```
Even now the scatterplot still seems to show discontinuities at the image centres.
```
from dials.array_family import flex
from matplotlib import pyplot
rt=flex.reflection_table.from_pickle('refined.pickle')

# remove unindexed reflections
rt=rt.select(rt['miller_index'] != (0,0,0))

# plot
xo,yo,zo=rt['xyzobs.px.value'].parts()
xc,yc,zc=rt['xyzcal.px'].parts()
zd = zo-zc
pyplot.scatter(zo, zd)
pyplot.xlabel('Zobs')
pyplot.ylabel('Zobs - Zcalc')
pyplot.ylim((-2,2))
pyplot.savefig('scatter.png')
```
![scatter](https://user-images.githubusercontent.com/11502083/42160245-67207eb2-7dee-11e8-9f44-2e61a1096d7b.png)

The centroid difference plot after `dials.report refined_experiments.json refined.pickle` also still looks odd. The sawtooth sections from the previous plots have gone, but what's left are centroids that still cluster around integer image numbers
![sawtooth5](https://user-images.githubusercontent.com/11502083/42160273-84293b52-7dee-11e8-96f1-60048d4b1172.png)

At this stage, I am not sure what to conclude. There might be a bug, but it's not even clear what it is. Does anybody have any ideas?

**Comment by rjgildea** [comment]
Reflections shouldn't be flagged as indexed if they have the index 0,0,0... are you saying this is not the case? I.e. reflections with the index 0,0,0 have the indexed flag set in the reflection table?

**Comment by biochem-fan** [comment]
Hi David, if the accuracy of Z centroids is the problem, what happens if centroids from integration are recycled? The 'true' signal might extend over more than one image, but spot finding might not be able to catch such tails. Can profile models and better background subtraction in integration improve it?

TPT datasets (https://data.sbgrid.org/dataset/512/) are small-wedge but fine-sliced (0.1 deg/frame). Does this problem happen there as well? Although the dataset is big, you can stop rsync after few datasets.

**Comment by dagewa** [comment]
No, I'm not saying that. Removing the unindexed reflections was only for making a scatterplot to explore the issue. I could have (better, should have) filtered using the `indexed` flag. The issue here is the systematic structure to the Z centroid residuals.

**Comment by dagewa** [comment]
I'll look at TPT datasets, but I'll start by seeing what happens with a normal, wide dataset when I only import the first few images.

**Comment by dagewa** [comment]
With the [I04 thaumatin training data](https://zenodo.org/record/10271#.WzoZhXYzrmE), it is probably unsurprising that the effect is always there, but it's not something I'd noticed before, typically looking at wider scans.
**Images 1-5:**
![newplot-01-05](https://user-images.githubusercontent.com/11502083/42163724-817862c2-7dfb-11e8-9302-22d404d29089.png)
**Images 1-9:**
![newplot-01-09](https://user-images.githubusercontent.com/11502083/42163730-87847dea-7dfb-11e8-916a-79315652f1b1.png)
**Images 1-30:**
![newplot-01-30](https://user-images.githubusercontent.com/11502083/42163740-8e749068-7dfb-11e8-8ee2-16a34b61233b.png)
It's a good idea to try recycling the centroids.

**Comment by rjgildea** [comment]
OK, I obviously misinterpreted the comment "Helps to remove the reflections indexed (0,0,0)"... reflections have only been "indexed" if the indexed flag has been set in the reflection table.

**Comment by dagewa** [comment]
The "sawtooth" plots now have their own issue #591, so this one should only be concerned with poor centroids at scan edges. The [trim-scan-edge-centroids](https://github.com/dials/dials/tree/trim-scan-edge-centroids) branch contains a fix that removes the worst centroids by rejecting anything within 0.1° of the centre of the first and last image in the scan. Ideally, rather than always picking 0.1°, the amount to trim should be chosen automatically in a way related to the rocking curve of reflections (note for later: perhaps the median `zsize()` of shoeboxes in the first and last degree of the scan would be a useful heuristic).

This fix is mainly intended for small wedge datasets, where the poor centroids at edges have the greatest effect. However, one issue is that we are throwing reflections away from refinement from datasets that don't have many reflections to start with. If there are too few reflections, refinement may then give poor results, or fail.

As a test, I tried processing some of the small wedge datasets referenced earlier. I downloaded all of the `unipuck2` datasets from https://zenodo.org/record/887440#.Wz9zmnYzrA0 and processed them with a simple DIALS script:
```
dials.import $1
dials.find_spots datablock.json
dials.index datablock.json strong.pickle unit_cell=55.9,65.0,160.7,90,90,90 \
  space_group=P212121
dials.refine experiments.json indexed.pickle
dials.integrate refined_experiments.json refined.pickle
dials.export integrated_experiments.json integrated.pickle
```
I did this both for the `master` and `trim-scan-edge-centroids` branches. Out of 240 datasets, I found:
- 170 datasets integrated on `master`
- 164 datasets integrated on `trim-scan-edge-centroids`
- 14 datasets integrated on `master` but not `trim-scan-edge-centroids`
- 8 datasets integrated on `trim-scan-edge-centroids` but not `master`

I have not gone through all the differences yet to see why some failed on one branch rather than the other, but in some cases where `master` succeeds where `trim-scan-edge-centroids` fails, it appears to be because there are too few reflections on the `trim` branch for indexing.

The RMSDs _should_ be better for `trim-scan-edge-centroids` compared to `master`, because we are specifically removing reflections we know to have poor observed centroids. That is indeed borne out by the refinement results (apologies for poor graphical presentation):
![rmsdx](https://user-images.githubusercontent.com/11502083/42382751-2a8a1728-812d-11e8-8d08-a41a3237d0af.png)
![rmsdy](https://user-images.githubusercontent.com/11502083/42382753-2ab4b7c6-812d-11e8-92ca-7d32bf286c36.png)
![rmsdz](https://user-images.githubusercontent.com/11502083/42382754-2aca406e-812d-11e8-909a-e5d74481e772.png)

I don't yet know how the `trim` branch has affected integration quality or merging statistics. That would be the next thing to look at.

**Comment by biochem-fan** [comment]
Hi David, thank you very much for detailed report. This looks promising. Joint refinement of detector and beam might also help. Keitaro (@keitaroyam), do you know how many crystals were integrated by XDS?

**Comment by biochem-fan** [comment]
Yes. I noticed instability of the detector distance in small wedge datasets like this. This is probably a similar issue to that in MicroED you mentioned in Acta D.

**Comment by graeme-winter** [comment]
@biochem-fan @dagewa I believe this to be false plotting in dials.report - see https://github.com/dials/dials/issues/594 for details - in the ticket @dagewa created looking at the "bag training" data other analysis indicated that the residuals are much better behaved than this plot would suggest.

**Comment by dagewa** [comment]
I'm now back to looking at this again after being side-tracked by other things for a while. Processing the `unipuck2` datasets with detector distance fixed gave:
- 178 datasets integrated on `master`
- 177 datasets integrated on `trim-scan-edge-centroids` 
- 6 integrated on `master` but not `trim-scan-edge-centroids`
- 5 integrated on `trim-scan-edge-centroids` but not `master`
The RMSDs after refinement compared:
![rmsdx](https://user-images.githubusercontent.com/11502083/42774193-03f9b5ba-8929-11e8-9ec6-68510c24fdbd.png)
![rmsdy](https://user-images.githubusercontent.com/11502083/42774192-03dcbe42-8929-11e8-8e1a-3eba2022826d.png)
![rmsdz](https://user-images.githubusercontent.com/11502083/42774191-03b8b2f4-8929-11e8-8455-607aaef748a3.png)
As expected, RMSDs are generally a little better for the `trim-scan-edge-centroids` branch. In two cases where `master` succeeded whilst `trim-scan-edge-centroids` failed the RMSDs are reasonably low. I should look at these cases to see why the branch is failing. The other cases have high RMSD and are more likely to be poor datasets. There are some other anomalies, like the 12th dataset. This integrates on both branches, but dials.export fails on the `trim-scan-edge-centroids` branch because there are no profile fitted reflections (falling foul of #555). I'll spend a bit of time investigating these corner cases before moving on to a comparison of integrated or merged data quality.

**Comment by dagewa** [comment]
It turns out that comparing merged data quality between the branches is complicated, so I started by comparing integration results (unscaled) between equivalent datasets. I adjusted the processing script to include a sensible resolution cutoff, so each dataset was processed on each branch using the following DIALS commands:
```
#!/bin/bash
echo $1
dials.import $1
dials.find_spots datablock.json filter.d_min=3.0
dials.index datablock.json strong.pickle unit_cell=55.9,65.0,160.7,90,90,90 \
  space_group=P212121 detector.fix_list=Dist
dials.refine experiments.json indexed.pickle output.history=history.pickle \
  detector.fix_list=Dist
dials.integrate refined_experiments.json refined.pickle prediction.d_min=3.0
dials.export integrated_experiments.json integrated.pickle
```
This produces 177 datasets that succeed in integration on both branches. I looked at the mean I/sigI in resolution shells for both profile fitting and summation integration. As each dataset is small, I used just 4 shells spanning 16 Å to 4 Å, equally sized in 1/d^2, which gave about 100-250 reflections in each bin. For every dataset I calculated the mean I/sigI in each bin, then looked at the difference between <I/sigI> between the `trim-scan-edge-centroids` and the `master` branch. This gives 177 differences in each resolution shell, which I then plotted:
![plot1](https://user-images.githubusercontent.com/11502083/43194948-92228208-8ffb-11e8-8954-2ff08a3df759.png)
![plot3](https://user-images.githubusercontent.com/11502083/43194961-9812bd04-8ffb-11e8-9808-acc036d1976e.png)
The differences are defined as the results from the `trim` branch minus the results from the `master` branch, so a tendency towards negative values indicates higher I/sigI on the master branch. It is clear that for both profile fitting and summation integration, in each resolution shell, the results from the `master` branch appear slightly better by this metric than results from the `trim` branch. So, despite a slight improvement in refinement RMSDs for almost all datasets on the `trim` branch, it seems that integration results are worse. This might be because fewer reflections are used in refinement on the `trim` branch. It may be that the reflections at the scan edge are useful for refinement for small wedge datasets despite having worse centroids.

Given these results, I won't merge the `trim-scan-edge-centroids` branch into `master`. However, the option to remove reflections at the scan edges may still be useful in other circumstances (perhaps for scan-varying refinement, with wider datasets). Therefore I'll leave the option available, but set by default to not remove any reflections.

During this investigation I noticed various problems with processing small wedge datasets with DIALS that are not related to refinement. @biochem-fan, perhaps some of these issues are behind your observation that XDS does better than DIALS in such cases. Do you know what XDS does when the mosaicity is a significant fraction of the dataset width? In DIALS, if no reflections are determined to be  fully recorded, then no profile models are made, so no profile fitting is done. Then following with standard use of `dials.export`, no MTZ will be written out at all. This seems sub-optimal, but we have identified a number of issues such as this and are dealing with them elsewhere (e.g. https://github.com/dials/dials/projects/8)

**Comment by dagewa** [comment]
This feature is now merged in master, so closing the issue

**Comment by biochem-fan** [comment]
@dagewa Thank you very much for detailed analysis!

> During this investigation I noticed various problems with processing small wedge datasets with DIALS that are not related to refinement. @biochem-fan, perhaps some of these issues are behind your observation that XDS does better than DIALS in such cases.

Yes.

> Do you know what XDS does when the mosaicity is a significant fraction of the dataset width? In DIALS, if no reflections are determined to be fully recorded, then no profile models are made, so no profile fitting is done.

I don't know the details but XDS can generate 3D profiles from a very narrow wedge, or even from a single image. See https://github.com/dials/dials/issues/473. According to the paper by @keitaroyam cited there, this feature does improve the data quality. 

I discussed a bit about this with @jmp1985, but we don't know how to proceed yet. Since more and more people are collecting small wedge datasets from microcrystals and/or in situ plates, improvements in this area would be beneficial to many users.

**Comment by dagewa** [comment]
Hi @biochem-fan, I hadn't noticed #473 before, very interesting. I agree this sounds important for small wedge data. I agree we should think about ways to allow profile fitting for these cases.
